### PR TITLE
💾 Feat/1142 Exception Manifest Generate Records

### PR DESCRIPTION
**Pull request by demariadaniel**
## Link to Issue

[#1142](https://app.zenhub.com/workspaces/icgc-argo-platform-dk-production-board-5e542d38415f5034e9fed89d/issues/gh/icgc-argo/argo-clinical/1142)

## Description

Collects requested Exception records filtered by ProgramId, DonorId, and Donor Submitter Ids

Unit Test output in ticket -- shows 5 types of formatted Exception records
Records are sorted according to spec but sorting is not enforced, will be completed in follow up ticket: #1143 

### Type of Change

- [ ] Bug
- [ ] Refactor
- [x] New Feature
- [ ] Release Candidate

### Checklist before requesting review:

- [x] Check branch (code change PRs go to `develop` not master)
- [x] Check copyrights for new files
- [x] Manual testing
- [x] Regression tests completed and passing (double check number of tests).
- [x] Spelling has been checked.
- [ ] Updated swagger docs accordingly (check it's still valid)
- [ ] Set `validationDependency` in meta tag for [Argo Dictionary](https://github.com/icgc-argo/argo-dictionary) fields used in code

### CODE REVIEW

File: src/exception/exception-manifest/util.ts
Comment (demariadaniel) [inline src/exception/exception-manifest/util.ts:35]: These 3 functions could likely be combined into a single generic function, writing this way just the path of least resistance

File: src/exception/exception-manifest/util.ts
Comment (demariadaniel) [inline src/exception/exception-manifest/util.ts:107]: Related to above `getEntityId` comment, this block could likely be improved

File: src/exception/exception-manifest/util.ts
Comment (demariadaniel) [reply src/exception/exception-manifest/util.ts]: Updated to use a generic function, see https://github.com/icgc-argo/argo-clinical/pull/1153/commits/fcc95d8a79ceb0c0d78d868d82470f24a2597b80 to compare with the previous version

File: src/exception/property-exceptions/util.ts
Comment (ciaranschutte) [inline src/exception/property-exceptions/util.ts:35]: curious how well this was working before, if this change is happening now.
I guess `treatment` was not working with the `isEntityException` function

File: package.json
Comment (demariadaniel) [reply package.json:13]: OK great, my first attempt was actually more specific and reduced # of tests run

Comment (demariadaniel) [reply package.json:13]: FYI this was impacting missing-entity-exception tests as well

File: src/exception/property-exceptions/util.ts
Comment (demariadaniel) [reply src/exception/property-exceptions/util.ts:35]: No appears it was missing

File: src/exception/exception-manifest/util.ts
Comment (ciaranschutte) [inline src/exception/exception-manifest/util.ts:62]: use constants available in codebase `schema === constant.specimen`
they're are multiple and duplicates, will need to see which fits your use case best

File: src/submission/exceptions/exceptions.ts
Comment (ciaranschutte) [inline src/submission/exceptions/exceptions.ts:309]: I'm unclear here what the difference between an `entityException` and a `missingEntityException` is.
There may be naming to tweak overall and this work is bringing out that improvement.
For here even some comments would help.

for `queryForExceptions` and `getByProgramId` it may be beneficial to import an object and then call the method, so more context is given, something like: `	const missingEntityException = await missingExceptionAPI.getByProgramId(programId);`

File: src/exception/exception-manifest/util.ts
Comment (ciaranschutte) [inline src/exception/exception-manifest/util.ts:85]: could rename this to `createProgramException` - it's not actually mapping anything.
this way when you use it it's self documenting: `[].map(createProgramException)`

File: src/exception/exception-manifest/util.ts
Comment (ciaranschutte) [inline src/exception/exception-manifest/util.ts:52]: this is exactly the same shape as  `entityKeyMap`
the type `entityKeyMap` is strictly enforcing a singular string not type of string
so these two can be changed to be a const and then get the type from it

File: src/exception/exception-manifest/util.ts
Comment (ciaranschutte) [inline src/exception/exception-manifest/util.ts:56]: really appreciate the individual function, helps the code a lot thank you 👍

File: src/exception/exception-manifest/util.ts
Comment (ciaranschutte) [inline src/exception/exception-manifest/util.ts:71]: this typeof check probably isn't needed

File: src/submission/exceptions/exceptions.ts
Comment (ciaranschutte) [inline src/submission/exceptions/exceptions.ts:296]: could add a single repo/mongo/db function somewhere to "find donor by submitter IDs (plural)", then there's no need for mapping through an individual query. possibility to even put some of the filtering there too.

File: src/exception/exception-manifest/util.ts
Comment (ciaranschutte) [inline src/exception/exception-manifest/util.ts:52]: these typings need to relate to the objects you're using.
there's a difference between 
```ts
type SpecimenEntityKey = 'specimenId'
```
and
```ts
type SpecimenEntityKey = Specimen['specimenId']
```

File: src/exception/exception-manifest/util.ts
Comment (ciaranschutte) [inline src/exception/exception-manifest/util.ts:80]: `Number(..)` I know why this is added but I don't think we should be forcing types like this anywhere.
It's similar to the the bug I just fixed where `JSON.stringify` was used to force types to be happy.
```ts
// this can technically be undefined
const value = clinicalRecord[entityKey];
// so Number(value) could be NaN
```

So there are 2 issues:
1 - DeepReadonly, we work around this one because we have not untangled it. it's everywhere in the codebase. it's a lot of non trivial work.
2 - Actual type issues. Everything that isn't deepreadonly. We either need to improve types in TS or we need to do the actual value checks in JS and assign a value appropriately. Simply casting variables is not a fix.

**Review by ciaranschutte — CHANGES_REQUESTED** [review]
Looks really great. I really appreciate using single functions to break the code into logically pieces. helps immensely. thank you.

Lots of suggestions/comments.

There's some TS usage we should fix before merging. cc @joneubank for TS feedback. in case there is something we could do in relation to deep read only. Maybe there's a stop gap to at least not add new code with it.

File: src/exception/exception-manifest/util.ts
Comment (demariadaniel) [reply src/exception/exception-manifest/util.ts]: Removed in [15ae5ee](https://github.com/icgc-argo/argo-clinical/pull/1153/commits/15ae5eec0b61a5d3ae8f61b0224191d8b616eecd)

File: src/exception/exception-manifest/util.ts
Comment (demariadaniel) [reply src/exception/exception-manifest/util.ts]: Updated in [15ae5ee](https://github.com/icgc-argo/argo-clinical/pull/1153/commits/15ae5eec0b61a5d3ae8f61b0224191d8b616eecd)

File: src/exception/exception-manifest/types.ts
Comment (demariadaniel) [reply src/exception/exception-manifest/types.ts]: I think I understand what you're saying here, updated in [15ae5ee](https://github.com/icgc-argo/argo-clinical/pull/1153/commits/15ae5eec0b61a5d3ae8f61b0224191d8b616eecd) but worth a second look to confirm

File: src/submission/exceptions/exceptions.ts
Comment (demariadaniel) [reply src/submission/exceptions/exceptions.ts:309]: We can alias this as `entityPropertyException`

File: src/submission/exceptions/exceptions.ts
Comment (demariadaniel) [reply src/submission/exceptions/exceptions.ts:296]: Yes! Added. Was surprised such a function didn't exist and now happy it was very easy to create.

File: src/exception/exception-manifest/util.ts
Comment (demariadaniel) [reply src/exception/exception-manifest/util.ts]: Updated

File: test/unit/exception/manifest/exceptionManifest.spec.ts
Comment (demariadaniel) [reply test/unit/exception/manifest/exceptionManifest.spec.ts:131]: Both updated

**Comment by ciaranschutte** [comment]
re-requesting review from myself as PR is ready for review based on slack message

File: src/exception/exception-manifest/types.ts
Comment (ciaranschutte) [inline src/exception/exception-manifest/types.ts:27]: Not requesting a change. Just adding 2cents.

If you structure these as an object like:
```ts
export const ExceptionType = {
	missingEntity: 'MissingEntity',
	entity: 'EntityProperty',
	program: 'ProgramProperty',
}
```
This means everywhere you manually put the string 'MissingEntity' you can do `ExceptionType.missingEntity`.
Then if you ever need to change it, you can change it in this object alone, and there will be no dangling references to a previous string.

Now we build some types to use throughout....

currently this is valid: `ExceptionType.missingEntity = 'the cake is a lie';`
now we've messed up our object.
this is also valid: `ExceptionType.fakeEntityForTesting = 123`
also messed up our object
How do we lock this down to just the values we declare originally? `as const`

```ts
export const ExceptionType = {
	missingEntity: 'MissingEntity',
	entity: 'EntityProperty',
	program: 'ProgramProperty',
} as const
```
by using `as const` we narrow the object.
`ExceptionType.missingEntity = 'the cake is a lie';` this doesnt work. the property can only be 'MissingEntity'.
we also can't add values to the object.

File: src/submission/exceptions/exceptions.ts
Comment (ciaranschutte) [inline src/submission/exceptions/exceptions.ts:340]: not a homogeneous array meaning it's holding different types of objects which I imagine being an issue here: https://github.com/icgc-argo/argo-clinical/pull/1161/files#diff-654b5ddd8e9c70e984a811687ea0784f6467af540e4cc7eb7ff9a831551e7cf7R75 when you try and make a TSV out of it.   (I changed the base branch to this branch so the diff looks right)

I think you'll end up with the union of all columns, which some records won't have any data for.
probably need to return an array of all the same type or an object with the difference types might be work too
```ts
const donorExceptionRecords = {programExceptions: [], entityExceptiosn: []}
```

File: test/unit/exception/manifest/exceptionManifest.spec.ts
Comment (ciaranschutte) [reply test/unit/exception/manifest/exceptionManifest.spec.ts:131]: this is great!

File: test/unit/exception/manifest/stubs.ts
Comment (demariadaniel) [reply test/unit/exception/manifest/stubs.ts:238]: Updated current testing, more updates in next PR

File: src/exception/exception-manifest/util.ts
Comment (ciaranschutte) [inline src/exception/exception-manifest/util.ts:55]: Not a requested change just some useful TS.

You can generate types from code.
```ts

// type EntityKeys = {
// 	[ClinicalEntitySchemaNames.SPECIMEN]: 'specimenId';
// 	[ClinicalEntitySchemaNames.TREATMENT]: 'treatmentId';
// 	[ClinicalEntitySchemaNames.FOLLOW_UP]: 'followUpId';
// };

const entityKeys = {
	[ClinicalEntitySchemaNames.SPECIMEN]: 'specimenId',
	[ClinicalEntitySchemaNames.TREATMENT]: 'treatmentId',
	[ClinicalEntitySchemaNames.FOLLOW_UP]: 'followUpId',
} as const;

type EntityKeys = typeof entityKeys;
```

**Review by ciaranschutte — APPROVED** [review]
approving! no need to hold this up.

one point about the return value that will go to the TSV download which I'd appreciate you taking a look at. can be addressed in future PR

everything else is just thoughts. no need to make changes.

File: src/submission/exceptions/exceptions.ts
Comment (demariadaniel) [reply src/submission/exceptions/exceptions.ts:340]: Yes this is correct, however we are only making 1 TSV for all exceptions, so I think we're going to see all columns regardless. I will confirm what's the expected TSV output!

File: src/exception/exception-manifest/types.ts
Comment (demariadaniel) [reply src/exception/exception-manifest/types.ts:27]: Excellent thank you, I will carry this over to the next PR

**Pull request merged**
